src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala

```diff
@@ -172,11 +178,14 @@ object RuleUtils {
     // Check pre-requisite.
     assert(getLogicalRelation(plan).isDefined)
 
-    val transformed = if (HyperspaceConf.hybridScanEnabled(spark)) {
-      transformPlanToUseHybridScan(spark, index, plan, useBucketSpec)
-    } else {
-      transformPlanToUseIndexOnlyScan(spark, index, plan, useBucketSpec)
-    }
+    val transformed =
+      if (HyperspaceConf.hybridScanEnabled(spark) && index
+        .getTagValue(IndexConstants.INDEX_HYBRIDSCAN_REQUIRED_TAG)
```

Comment:
Hmm. To me, this seems a bit fragile since this is calculated in `getCandidateIndexes` and used in `transformPlanToUseIndex`. How can we guarantee that the tag was generated with the correct `plan`?

Reply:
getCandidateIndex(plan) creates copy of indexLogEntry - so, for example, JoinIndexRule has `lIndex` and `rIndex` for left plan and right plan. Candidates indexes for both are selected and applied respectively. In this way, we could keep the tags for each plan.

Reply:
More example in rank:
https://github.com/microsoft/hyperspace/pull/164/files#diff-57915e1583c4c52ae6ccbf6324db8db200813f3c8bf3b6b4623b299ca80aaa64L44

Reply:
Can't you just tag the logical plan used? I am just afraid that it may become harder to maintain in a long run.

Reply:
There's no tag in Spark 2.4... that's the problem. 😟

Reply:
I mean tagging the plan to the log entry?

Reply:
Ok then I'll try sth like
```
private val tags: mutable.Map[IndexLogEntryTag[_], (plan, Any)] = mutable.Map.empty
```

Reply:
Can't you have `IndexLogEntryTag[LogicalPlan]("plan")` as a separate key? And make sure you are operating on the same plan?

Reply:
How about this
```
@JsonIgnore
  private val tags: mutable.Map[(LogicalPlan, IndexLogEntryTag[_]), Any] = mutable.Map.empty

  def setTagValue[T](plan: LogicalPlan, tag: IndexLogEntryTag[T], value: T): Unit = {
    tags((plan, tag)) = value
  }

  def getTagValue[T](plan: LogicalPlan, tag: IndexLogEntryTag[T]): Option[T] = {
    tags.get((plan, tag)).map(_.asInstanceOf[T])
  }

  def unsetTagValue[T](plan: LogicalPlan, tag: IndexLogEntryTag[T]): Unit = {
    tags.remove((plan, tag))
  }
}

// A tag of a `IndexLogEntry`, which defines name and type.
case class IndexLogEntryTag[T](name: String)
```

Reply:
Yea, this seems fine. And I guess we could use `null` if a tag is not associated with a plan?

```
def setTagValue[T](tag: IndexLogEntryTag[T], value: T): Unit = {
  tags((null, tag)) = value
}
```

Reply:
Comment test 4